### PR TITLE
Refactor `LockedAmplitude` and `ShapedAmplitude`

### PR DESCRIPTION
**Pull request by goerz**
Both of these amplitudes could be made more flexible by getting rid of the type hierarchy that distinguished between, e.g., `ShapedContinousAmplitude` and `ShapedPulseAmplitude`. The shape and control can now be vectors or functions independently of each other. This enables substituting optimized pulses into existing amplitudes, among other benefits.

### CODE REVIEW

File: test/test_amplitudes.jl [hunk 1]
```diff
@@ -25,6 +25,7 @@ using QuantumPropagators.Controls: discretize_on_midpoints # DEBUG
     @test length(get_controls(ampl)) == 0
     t = t_mid(tlist, 20)
     @test evaluate(ampl, tlist, 20) ≈ S(t)
+    @test_throws Exception evaluate(ampl, t)
```
[the remark this answers is not in the document]
Comment (goerz) [reply test/test_amplitudes.jl]: Yeah, good point. I'll use `IOCapture` for this

File: src/amplitudes.jl [hunk 2]
```diff
@@ -15,29 +15,34 @@ using ..Controls: t_mid
...
+            end
+        end
+        return new{ST}(shape)
     end
 end
 
 
 function LockedAmplitude(shape, tlist)
-    return LockedPulseAmplitude(discretize_on_midpoints(shape, tlist))
+    return LockedAmplitude(discretize_on_midpoints(shape, tlist); check = false)
```
[the remark this answers is not in the document]
Comment (goerz) [reply src/amplitudes.jl:57]: We'll fix `discretize_on_midpoints` to always return `Vector{Float64}`

File: src/amplitudes.jl [hunk 3]
```diff
@@ -124,63 +105,122 @@ ampl = ShapedAmplitude(control; shape=shape)
...
     print(io, "ShapedAmplitude(::$(typeof(ampl.control)); shape::$(typeof(ampl.shape)))")
 end
 
-function ShapedAmplitude(control, tlist; shape)
+
+function ShapedAmplitude(control, tlist::Vector{Float64}; shape)
     control = discretize_on_midpoints(control, tlist)
     shape = discretize_on_midpoints(shape, tlist)
-    return ShapedPulseAmplitude(control, shape)
+    return ShapedAmplitude{typeof(control),typeof(shape)}(control, shape)
```
[the remark this answers is not in the document]
Comment (goerz) [reply src/amplitudes.jl]: Yeah, we need to normalize the type

File: src/amplitudes.jl [hunk 4]
```diff
@@ -124,63 +105,122 @@ ampl = ShapedAmplitude(control; shape=shape)
...
+    return ShapedAmplitude(control; shape = ampl.shape, check = true)
+end
+
+
+Base.Array(ampl::ShapedAmplitude{Vector{Float64},Vector{Float64}}) =
+    ampl.control .* ampl.shape
+
+(ampl::ShapedAmplitude)(t::Float64) = ampl.shape(t) * ampl.control(t)
+
+
```
[the remark this answers is not in the document]
Comment (goerz) [reply src/amplitudes.jl:206]: I don't think that's worth the complication. Should be an edge case

File: test/test_amplitudes.jl [hunk 5]
```diff
@@ -57,5 +58,54 @@ end
...
+
+end
+
+
+@testset "ShapedAmplitude mixed" begin
+    tlist = collect(range(0, 10, length = 101))
+    S(t) = flattop(t, T = 10, t_rise = 2, func = :blackman)
+    ϵ(t) = 1.0
+    S_vec = discretize_on_midpoints(S, tlist)
+    ϵ_vec = discretize_on_midpoints(ϵ, tlist)
```
[the remark this answers is not in the document]
Comment (goerz) [reply test/test_amplitudes.jl:120]: We can enforce everything to be `Vector{Float64}`
